bytecomp/translmod.ml

```diff
@@ -1358,27 +1377,41 @@ let transl_store_package component_names target_name coercion =
 
 open Format
 
-let print_cycle ppf =
-  Format.pp_print_list ~pp_sep:(fun ppf () -> fprintf ppf "@ -> ")
-    Printtyp.ident ppf
-
-let report_error ppf = function
-    Circular_dependency cycle ->
+let print_cycle ppf cycle =
+  let print_ident ppf (x,_) = Format.pp_print_string ppf (Ident.name x) in
+  let pp_sep ppf () = fprintf ppf "@ -> " in
+  Format.fprintf ppf "%a%a%s"
+    (Format.pp_print_list ~pp_sep print_ident) cycle
+    pp_sep ()
+    (Ident.name @@ fst @@ List.hd cycle)
+(* we repeat the first element to make the cycle more apparent *)
+
+let explanation_submsg (id, {reason;loc;subid}) =
+  let print fmt =
+    let printer = Format.dprintf fmt (Ident.name id) (Ident.name subid) in
+    Location.mkloc printer loc in
+  match reason with
+  | Unsafe_module_binding -> print "Module %s defines an unsafe module, %s ."
+  | Unsafe_functor -> print "Module %s defines an unsafe functor, %s ."
+  | Unsafe_typext ->
+      print "Module %s defines an unsafe extension constructor, %s ."
+  | Unsafe_non_function -> print "Module %s defines an unsafe value, %s ."
+
+let report_error loc = function
+  | Circular_dependency cycle ->
       let[@manual.ref "s-recursive-modules"] chapter, section = 8, 3 in
-      fprintf ppf
-        "@[Cannot safely evaluate the definition of the following cycle@ \
+      Location.errorf ~loc ~sub:(List.map explanation_submsg cycle)
+        "Cannot safely evaluate the definition of the following cycle@ \
          of recursively-defined modules:@ %a.@ \
-         There are no safe modules in this cycle@ (see manual section %d.%d)@]"
+         There are no safe modules in this cycle@ (see manual section %d.%d)."
         print_cycle cycle chapter section
   | Conflicting_inline_attributes ->
-      fprintf ppf
-        "@[Conflicting ``inline'' attributes@]"
+      Location.errorf "@[Conflicting ``inline'' attributes@]"
```

Comment:
Unrelated, but is there a reason to use backquotes and double-singlequotes here instead of just quotes?

Reply:
It feels very TeX. :)

Reply:
I don't know. @mshinwell , it looks like you are the one that wrote this error message in cd8a4a9c5d0db, do you remember any particular reason?

Reply:
There was no particular reason.  They should probably be singles.

Reply:
Done: I replaced ` ``inline'' ` by `'inline'`.